Compfiles/Poland2016S1P8.lean

```diff
@@ -0,0 +1,148 @@
+/-
+Copyright (c) 2023 The Compfiles Contributors. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Adam Kurkiewicz
+-/
+
+import Mathlib.Data.Nat.ModEq
+import Mathlib.Tactic
+
+import ProblemExtraction
+
+problem_file { tags := [.NumberTheory] }
+
+/-!
+
+Polish Mathematical Olympiad 2016, Stage 1, Problem 8
+Author of the problem: Nguyen Hung Son
+Source of the problem: https://om.sem.edu.pl/static/app_main/problems/om68_1r.pdf
+
+Let a, b, c be integers. Show that there exists a positive integer n, such that
+
+  n³ + an² + bn + c
+
+is not a square of any integer
+
+-/
+
+namespace Poland2016S1P8
+
+snip begin
+
+lemma even_of_add {a : ℤ} {b : ℤ} : Even a → Even (a + b) → Even (b) := by
+  intro A
+  obtain ⟨k, H⟩ := A
+  intro A_B
+  obtain ⟨k2, H2⟩ := A_B
+  use (k2 - k)
+  rw [H] at H2
+  apply symm
+  calc k2 - k + (k2 - k) = k2 + k2 - k - k := by ring
+    _ = k + k + b - k -k := by rw[H2]
+    _ = b := by ring
+
+lemma div_4_mul_of_both_even {a b : ℤ } (H : Even a ∧ Even b) : 4 ∣ a * b := by
+  obtain ⟨k, Hk⟩ := H.left
+  obtain ⟨l, Hl⟩ := H.right
+  use k * l
+  rw[Hk]
+  rw[Hl]
+  ring
+
+snip end
+
+def is_square (n : ℤ) : Prop :=
```

Comment:
Can you use mathlib's `IsSquare`? https://github.com/leanprover-community/mathlib4/blob/5e33eb3b841bfdcee4fff013790d29ce38df7056/Mathlib/Algebra/Group/Even.lean#L48-L48